core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java

```diff
@@ -651,109 +662,154 @@ public void close() {
     * @param <S>
     */
    private class CompletionListenerRehashIterator<S> extends RehashIterator<S> {
-      private final KeyWatchingCompletionListener completionListener;
+      private final Consumer<? super Supplier<PrimitiveIterator.OfInt>> userListener;
+      private final List<KeyWatchingCompletionListener> completionListeners;
 
       private CompletionListenerRehashIterator(Iterable<IntermediateOperation> intermediateOperations,
             Consumer<? super Supplier<PrimitiveIterator.OfInt>> completionListener) {
          super(intermediateOperations);
-         this.completionListener = new KeyWatchingCompletionListener(completionListener);
+         this.completionListeners = Collections.synchronizedList(new ArrayList<>(4));
+         this.userListener = completionListener;
       }
 
       @Override
       protected S getNext() {
          S next = super.getNext();
          if (next != null) {
-            completionListener.valueIterated(next);
+            for (KeyWatchingCompletionListener keyWatchingCompletionListener : completionListeners) {
```

Comment:
is it guaranteed that all listeners will be initialized when this gets called?

Reply:
Yes, they are always initialized the way the code works.

And actually if this changes it doesn't matter since the listeners only care about entries that were retrieved by their own publisher.